fix(atc): when secret redaction is not enabled, should not buffer build log lines

evanchaoli opened the pull request:
# Existing Issue

Fixes #4799 .

# Changes proposed in this pull request

When secret redaction is not enabled, use `newDBEventWriter` to create stdout and stderr.

# Contributor Checklist
> Are the following items included as part of this PR? Please delete checkbox items that don't apply.
- [ ] Unit tests
- [ ] Integration tests (if applicable)
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)


# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the [BOSH](https://github.com/concourse/concourse-bosh-release) 
      and [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for the [integration tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env) (for example, if they are Garden configs that are not displayed in the `--help` text).

evanchaoli commented:
@vito This PR, together with #4757 and #4808, maybe worthy a minor release of 5.6.1.

deniseyu commented:
Have to admit I don't have deep familiarity with this part of the code (but I wanted to build some context 😄) -- from reading the code it looks like @zoetian's observation in https://github.com/concourse/concourse/issues/4799 that redaction is always opted in is indeed the case, and switching the flag off [results in the DummyCredVarsTracker being used](https://github.com/concourse/concourse/pull/4817/commits/17eb00662e3c79e49ba4de272d58e6a9db2f4725#diff-68ac07b71ffa26113681f2aec23dca07R23-L30). Hard-coding the `Enabled()` return value as `true` further reinforces the idea that it's either redact or use the dummy.

I didn't review the last PR that merged this logic in so I could just be missing context, but it seems odd that if an operator disables this flag, then we send them down the path of using a dummy struct. But maybe it's just a naming issue, that the dummy struct isn't actually a placeholder or no-op object in the same sense that we use dummies in other parts of the codebase; in which case, we should rename it away from "dummy" and to something like "noRedact".

evanchaoli commented:
@deniseyu Let me explain the story a little bit. I implemented the secret redaction feature, and added a CLI option to switch the feature. The way to turn off secret redaction is to return `dummyCredVarsTracker` from `NewCredVarsTracker` when the second parameter `on` is false. The theory here is, `dummyCredVarsTracker` is no-op, if we don't track secret, then it won't redact any secret. 

The regression is in `buildStepDelegate`. Because `buildStepDelete` gets only an object of `CredVarsTracker`, it doesn't know if or not secret redaction is on or off, thus I always created `dbEventWriterWithSecretRedaction` object for `Stdout()` and `Stderr()`, which always buffers line.

With this change, I added `Enabled()` to interface `CredVarsTracker`, so that `buildStepDelegator` may know if var tracker is on or off, when off, use the old `dbEventWriter`, which doesn't buffer line, so that avoid regression.

If you concern with the name of `dummyCredVarsTracker`, I can rename it to `noOpTracker`. But I don't quite understand what did you mean by `to something like "noRedact" (which might also let us eliminate quite a bit of code in vars_tracker.go)`, can you please explain?

evanchaoli commented:
@pivotal-jwinters This PR is target to fix the regression. I think the issue #4799 is saying that even if they don't turn on secret redaction, build logs are still buffered. 

I agree that ideally we should resolve the issue by handling "\r" properly. We have seen a couple of use cases, #4749 and #4799, that my design of secret redaction doesn't cover. So I'd wait for more issue reports, and then think about a solution to try to cover all use cases.
